### DISCUSSION TITLE
Event buckets

### DISCUSSION THREAD

**Discussion by afrittoli** [cdfoundation/sig-events#44] [General]
We organised our protocol events in [four buckets](https://github.com/cdfoundation/sig-events/blob/main/vocabulary-draft/introduction.md) (or phases):

- Source code version control
- Continuous integration pipelines
- Continuous deployment pipelines
- Continuous delivery (or [core](https://github.com/cdfoundation/sig-events/blob/main/vocabulary-draft/core.md)) events

These buckets are a good way to organise our documentation and perhaps protocol spec.
We should clarify what the bucket mean from a protocol and implementor point of view.

The initial proposal @salaboy was to include the bucket name in the event name.

Advantages:
- it allows for filtering of a group of events with a regex / rule on the event name

Disadvantages:
- changing buckets, moving events between buckers has an impact on the protocol
- it only allows for one group for each event

An alternative solution could be to leave the bucket name outside of the event type, which would gives us more flexibility in changing the buckets without impacting the protocol directly.

We could allow for (optional?) multi-value custom headers to specify categories. This would allow to:
- still filter events based on header info
- allow for multiple categories (tags) for events

For instance an event could be in the `ci` bucket and related to `testing`, and that could be mapped to two tags. 
Perhaps in future we could decide that `testing` is a bucket of its own. 

```
CURL -X POST example/ce-body.json \
-H 'ce-id: 1' \
-H 'ce-specversion: 1.0' \
-H 'ce-type: dev.eicd.testing.started' \
-H 'ce-source: http://localhost' \
-H 'eicd-tag: ci' \
-H 'eicd-tag: testing'
```

**Comment by erkist** [comment]
> These buckets are a good way to organise our documentation and perhaps protocol spec.

This is one of the main values of buckets for me, to organize events. To make it easier for users starting in some area  to see the events they should care about or use.

This doesn't, however, actually need to be reflected in the event type, we could have a "flat list" of events from an event type perspective, and deal with grouping in the documentation.

I don't have an opinion on the use of tags, as I don't understand the use case of filtering events by group rather than by specific type. Is it for when "we only want to copy events of this specific group to this public message bus"? What if some old event is suddenly update to include a new tag? Seems risky.

**Comment by afrittoli** [comment]
In `keptn` services can subscribe to event types based on filters, which is a use cases for having group of events of some kind.
This could be achieved through dedicated HTTP headers.

**Comment by afrittoli** [comment]
Related issues: https://github.com/cdfoundation/sig-events/issues/45

**Comment by AloisReitbauer** [comment]
Having buckets would limit in which stage a certain task type is used. Testing - as an example - is used in CI, CD or progressive delivery or even production for chaos testing. Currently the buckets are heavily tied to lifecycle stages.

**Comment by m-linner-ericsson** [comment]
I agree with @AloisReitbauer and would prefer not to have the bucket in the events themselves. Tags is a much better way if we want some kind of grouping.
